flink-runtime/src/main/java/org/apache/flink/runtime/state/heap/HeapKeyedStateBackend.java

```diff
@@ -630,19 +506,210 @@ public int numStateEntries(Object namespace) {
+			final Map<String, Integer> kVStateToId = new HashMap<>(stateTables.size());
+
+			final Map<StateTable<K, ?, ?>, StateTableSnapshot> cowStateStableSnapshots =
+				new HashedMap(stateTables.size());
+
+			for (Map.Entry<String, StateTable<K, ?, ?>> kvState : stateTables.entrySet()) {
+				kVStateToId.put(kvState.getKey(), kVStateToId.size());
+				StateTable<K, ?, ?> stateTable = kvState.getValue();
+				if (null != stateTable) {
+					metaInfoSnapshots.add(stateTable.getMetaInfo().snapshot());
+					cowStateStableSnapshots.put(stateTable, stateTable.createSnapshot());
+				}
+			}
+
+			final KeyedBackendSerializationProxy<K> serializationProxy =
+				new KeyedBackendSerializationProxy<>(
+					keySerializer,
+					metaInfoSnapshots,
+					!Objects.equals(UncompressedStreamCompressionDecorator.INSTANCE, keyGroupCompressionDecorator));
+
+			//--------------------------------------------------- this becomes the end of sync part
+
+			// implementation of the async IO operation, based on FutureTask
+			final AbstractAsyncCallableWithResources<SnapshotResult<KeyedStateHandle>> ioCallable =
+				new AbstractAsyncCallableWithResources<SnapshotResult<KeyedStateHandle>>() {
+
+					CheckpointStreamFactory.CheckpointStateOutputStream stream = null;
+
+					@Override
+					protected void acquireResources() throws Exception {
+						stream = streamFactory.createCheckpointStateOutputStream(checkpointId, timestamp);
+						cancelStreamRegistry.registerCloseable(stream);
+					}
+
+					@Override
+					protected void releaseResources() throws Exception {
+
+						if (cancelStreamRegistry.unregisterCloseable(stream)) {
+							IOUtils.closeQuietly(stream);
+							stream = null;
+						}
+
+						for (StateTableSnapshot tableSnapshot : cowStateStableSnapshots.values()) {
+							tableSnapshot.release();
+						}
+					}
+
+					@Override
+					protected void stopOperation() throws Exception {
+						if (cancelStreamRegistry.unregisterCloseable(stream)) {
+							IOUtils.closeQuietly(stream);
+							stream = null;
+						}
+					}
+
+					@Override
+					protected SnapshotResult<KeyedStateHandle> performOperation() throws Exception {
+
+						long startTime = System.currentTimeMillis();
+
+						CheckpointStreamFactory.CheckpointStateOutputStream localStream = this.stream;
+
+						DataOutputViewStreamWrapper outView = new DataOutputViewStreamWrapper(localStream);
+						serializationProxy.write(outView);
+
+						long[] keyGroupRangeOffsets = new long[keyGroupRange.getNumberOfKeyGroups()];
+
+						for (int keyGroupPos = 0; keyGroupPos < keyGroupRange.getNumberOfKeyGroups(); ++keyGroupPos) {
+							int keyGroupId = keyGroupRange.getKeyGroupId(keyGroupPos);
+							keyGroupRangeOffsets[keyGroupPos] = localStream.getPos();
+							outView.writeInt(keyGroupId);
+
+							for (Map.Entry<String, StateTable<K, ?, ?>> kvState : stateTables.entrySet()) {
+								OutputStream kgCompressionOut = keyGroupCompressionDecorator.decorateWithCompression(localStream);
```

Comment:
Can we put this in a try with resource statement?

Reply:
👍